src/Libraries/Storage/Adapters/GoogleDrive/GoogleDriveFileSystemAdapter.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+namespace Quantum\Libraries\Storage\Adapters\GoogleDrive;
+
+use Quantum\Libraries\Storage\FilesystemAdapterInterface;
+use Exception;
+
+class GoogleDriveFileSystemAdapter implements FilesystemAdapterInterface
+{
+
+    /**
+     * @var GoogleDriveFileSystemAdapter|null
+     */
+    private static $instance = null;
+
+    private $googleDriveApp;
+
+    /**
+     * GoogleDriveFileSystemAdapter constructor
+     * @param GoogleDriveApp $googleDriveApp
+     */
+    private function __construct(GoogleDriveApp $googleDriveApp)
+    {
+        $this->googleDriveApp = $googleDriveApp;
+    }
+
+    /**
+     * Get Instance
+     * @param GoogleDriveApp $googleDriveApp
+     * @return GoogleDriveFileSystemAdapter
+     */
+    public static function getInstance(GoogleDriveApp $googleDriveApp): GoogleDriveFileSystemAdapter
+    {
+        if (self::$instance === null) {
+            self::$instance = new self($googleDriveApp);
+        }
+
+        return self::$instance;
+    }
+
+    public function makeDirectory($dirname, $parentId = null): bool
+    {
+        try{
+            $data = [
+                'name' => $dirname,
+                'mimeType' => GoogleDriveApp::FOLDER_MIMETYPE,
+                'parents' => $parentId ? [$parentId] : ['root']
+            ];
+
+            $this->googleDriveApp->rpcRequest(GoogleDriveApp::FILE_METADATA_URL, $data);
+            return true;
+        } catch (Exception $e) {
+            return false;
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function removeDirectory(string $dirname): bool
+    {
+        return $this->remove($dirname);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function get(string $filename)
+    {
+        try {
+            return (string)$this->googleDriveApp->getFileInfo($filename, true);
+        } catch (Exception $e) {
+            return false;
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function put(string $filename, string $content, string $parentId = null)
+    {
+        try {
+            if($this->isFile($filename)){
+                return $this->googleDriveApp->rpcRequest(GoogleDriveApp::FILE_MEDIA_URL . '/' . $filename . '?uploadType=media', $content, 'PATCH', 'application/octet-stream');
+            }else{
```

Comment:
Is this really necessary to do this, doesn't google create new file (if not exists) or overwrite the file (if exists) ?

Reply:
Google Drive itself doesn't support that, when overriding a file, you have to specify the file's id, whilst creating one requires a file name.